pydis_site/apps/api/serializers.py

```diff
@@ -141,30 +144,270 @@ class Meta:
+    'disabled_channels',
+    'enabled_categories',
+    'disabled_categories',
+)
+
+ALLOW_BLANK_SETTINGS = (
+    'dm_content',
+    'dm_embed',
+    'infraction_reason',
+)
+
+ALLOW_EMPTY_SETTINGS = (
+    'enabled_channels',
+    'disabled_channels',
+    'enabled_categories',
+    'disabled_categories',
+    'guild_pings',
+    'dm_pings',
+    'bypass_roles',
+)
+
+# Required fields for custom JSON representation purposes
+BASE_FILTER_FIELDS = (
+    'id', 'created_at', 'updated_at', 'content', 'description', 'additional_field'
+)
+BASE_FILTERLIST_FIELDS = ('id', 'created_at', 'updated_at', 'name', 'list_type')
+BASE_SETTINGS_FIELDS = (
+    "bypass_roles",
+    "filter_dm",
+    "enabled",
+    "remove_context",
+    "send_alert"
+)
+INFRACTION_AND_NOTIFICATION_FIELDS = (
+    "infraction_type",
+    "infraction_reason",
+    "infraction_duration",
+    "infraction_channel",
+    "dm_content",
+    "dm_embed"
+)
+CHANNEL_SCOPE_FIELDS = (
+    "disabled_channels",
+    "disabled_categories",
+    "enabled_channels",
+    "enabled_categories"
+)
+MENTIONS_FIELDS = ("guild_pings", "dm_pings")
+
+
+def _create_filter_meta_extra_kwargs() -> dict[str, dict[str, bool]]:
+    """Create the extra kwargs of the Filter serializer's Meta class."""
+    extra_kwargs = {}
+    for field in SETTINGS_FIELDS:
+        field_args = {'required': False, 'allow_null': True}
+        if field in ALLOW_BLANK_SETTINGS:
+            field_args['allow_blank'] = True
+        if field in ALLOW_EMPTY_SETTINGS:
+            field_args['allow_empty'] = True
+        extra_kwargs[field] = field_args
+    return extra_kwargs
+
+
+def get_field_value(data: dict, field_name: str) -> Any:
+    """Get the value directly from the key, or from the filter list if it's missing or is None."""
+    if data.get(field_name) is not None:
+        return data[field_name]
+    return getattr(data["filter_list"], field_name)
+
+
+class FilterSerializer(ModelSerializer):
+    """A class providing (de-)serialization of `Filter` instances."""
+
+    def validate(self, data: dict) -> dict:
+        """Perform infraction data + allowed and disallowed lists validation."""
+        if (
+            (
+                get_field_value(data, "infraction_reason")
+                or get_field_value(data, "infraction_duration")
+            )
+            and get_field_value(data, "infraction_type") == "NONE"
+        ):
+            raise ValidationError(
+                "Infraction type is required with infraction duration or reason."
+            )
+
+        if (
+            set(get_field_value(data, "disabled_channels"))
+            & set(get_field_value(data, "enabled_channels"))
+        ):
+            raise ValidationError(
+                "You can't have the same value in both enabled and disabled channels lists."
+            )
+
+        if (
+            set(get_field_value(data, "disabled_categories"))
+            & set(get_field_value(data, "enabled_categories"))
```

Comment:
I believe these should be single quotes.

Reply:
[Use consistent quoting style](https://github.com/python-discord/site/pull/861/commits/91b89475913210400cf39884efe37ab5552efbf7)